app/controllers/companies/products_controller.rb

```diff
@@ -3,5 +3,11 @@
 class Companies::ProductsController < ApplicationController
   def index
     @company = Company.find(params[:company_id])
+    @products = Product.joins(user: :company)
+                       .where('company_id = ?', @company.id)
```

Comment:
```suggestion
                       .where({ users: {company_id: @company.id } })
```

こんな感じでテーブルを指定できるはずです〜